nntrainer/models/neuralnet.cpp

```diff
@@ -194,10 +194,16 @@ void NeuralNetwork::backwarding(std::shared_ptr<LayerNode> node, int iteration,
                                 bool calc_derivative) {
   /**
    * Do not change this order:
+   * 0. tryZeroGrad
    * 1. calcGradient
    * 2. calcDerivative
+   * 3. tryApplyGradients
    */
-  [[maybe_unused]] bool apply_gradient = true;
+
+  /// try zeroing the gradient at the first shared weight access
+  model_graph.tryZeroGrad(node.get());
```

Comment:
add TODO that this should only be done for shared weights, and otherwise should not be done to avoid this computation overhead.

Reply:
This has to be done for everyone as gradients are now being added not overwritten.

Reply:
For now, maybe, but we don't want to do this for all the layers because this adds extra memory as well as computation overhead. This is only required for weights that are shared. This can be handled at the weight level itself without layers changing their implementation.

Reply:
As discussed offline, will check if this makes some significant overhead, else it will be done within this PR

Reply:
Updated as a new commit.